### ISSUE TITLE
POST /flows with Node-RED-Deployment-Type 'nodes' works incorrectly

### ISSUE THREAD

**Issue by piotrpad** [node-red/node-red#979]
According to http://nodered.org/docs/api/admin/methods/post/flows/ POSTing flow (identical to one running) with changed ONE node with Node-RED-Deployment-Type = 'nodes' should only stop/start this one node.
Unfortunately this is not the case. My flow has following structure: 
flow(node, node, node, subflow(nodeX), node)  (can post json if needed)
and when I make a POST request with data where ONLY changed node (namely: I use delay node and modify delay) is nodeX (the one inside subflow) the following seem to be stopped:
nodeX - this is correct, node changed
subflow - this is not correct, Node-RED-Deployment-Type = 'nodes' - only changed nodes should be stopped/started
flow - this is not correct, same as above

(I know that because I added some more logging into node-red/red/runtime/nodes/flows/index.js)
According to documentation it would be bug. Another thing is - how that would be different from same situation with Node-RED-Deployment-Type = 'flows'?

**Comment by knolleary** [comment]
This is a documentation issue rather than code issue.

If you modify anything inside a subflow template we consider the subflow instance node to have changed, so the whole subflow instance is stopped and restarted.

**Comment by piotrpad** [comment]
Ok, this seems clear, but why the "main" flow also gets restarted?

**Comment by piotrpad** [comment]
Probably a bit of a background what I want to achieve will help. Imagine a flow that gets data from "outside", each msg is tagged, say with "1". One of the first nodes in the flow is a switch which redirects msgs tagged with "1" to one part of flow and msgs tagged with "2" to another part of the flow. At the beginning all the messages have tag "1" and thus flow part "for" tag "2" is inactive (that is all nodes connected to second switch "output"). At some point we need to do some different processing on certain msgs. Problem is, we do not know a'priori (when the flow is initially defined) what processing that would be. For this I planned to do the following in chronological order:
a) update nodes that are "after" switch output 2 to do processing needed - during this, active part of the flow (input, switch and all nodes "after" switch output 1) should continuously work
b) start sending msgs tagged with "2" so the switch redirects them to output 2 for processing by freshly defined/updated nodes (in point a).

Crucial point is that I cannot afford loosing any msg. So I do not want the "main" flow to be stopped.

**Comment by piotrpad** [comment]
And last question, already present in initial description - if in 'nodes' mode node and all corresponding flows are restarted how that is different from 'flows' update mode?

**Comment by knolleary** [comment]
Hi @piotrpad,

in my testing I cannot get it to fail in the way you describe.

Below is the flow I'm using which I think reflects what you describe your flow as.

    Inject -> Function -> Subflow [ Function ] -> Debug

The inject node fires every second. The first Function node prepends a count that increases with each message, as does the Function inside the Subflow.

So when running, the debug shows:

    1 : 1 : 1473366520568
    2 : 2 : 1473366521571
    3 : 3 : 1473366522571
    4 : 4 : 1473366523573
    ....

The first number from the Subflow Function, the second from the first Function.

If I do a Full deploy, both of the Function nodes are restarted as expected so their counts reset to 1.

If I modify the Subflow Function node and do a Modified Flows deploy, again both counts reset as expected.

If I modify the Subflow Function node and do a Modified Nodes deploy, only the first count resets to 1. This indicates the other nodes have been left running.



    [{"id":"9b4d6332.e3362","type":"subflow","name":"Subflow 1","info":"","in":[{"x":390,"y":120,"wires":[{"id":"1833aa51.eb6e96"}]}],"out":[{"x":650,"y":120,"wires":[{"id":"1833aa51.eb6e96","port":0}]}]},{"id":"1833aa51.eb6e96","type":"function","z":"9b4d6332.e3362","name":"","func":"var count = (context.get('count')||0)+1;\ncontext.set('count',count);\nmsg.payload = count+\" : \"+msg.payload;\nreturn msg;","outputs":1,"noerr":0,"x":510,"y":120,"wires":[[]]},{"id":"10c88f6a.168191","type":"inject","z":"64dbd79c.6b7788","name":"","topic":"","payload":"","payloadType":"date","repeat":"1","crontab":"","once":false,"x":150,"y":80,"wires":[["1e397859.bb8ea8"]]},{"id":"1e397859.bb8ea8","type":"function","z":"64dbd79c.6b7788","name":"","func":"var count = (context.get('count')||0)+1;\ncontext.set('count',count);\nmsg.payload = count+\" : \"+msg.payload;\nreturn msg;","outputs":1,"noerr":0,"x":350,"y":80,"wires":[["f897a10f.5ca9e"]]},{"id":"20019a6d.e953d6","type":"debug","z":"64dbd79c.6b7788","name":"","active":true,"console":"false","complete":"false","x":750,"y":80,"wires":[]},{"id":"f897a10f.5ca9e","type":"subflow:9b4d6332.e3362","z":"64dbd79c.6b7788","name":"","x":540,"y":80,"wires":[["20019a6d.e953d6"]]}]

**Comment by piotrpad** [comment]
Hi @knolleary,
unfortunately I'm not able to reproduce this behavior. In my setup context 'count' variables seem not to be reset at all - at least in cases you mention (even full redeploy after change in the node does not reset them). 

things to clarify:
- to make things more simple I used node-red GUI instead of "artificial" POST request from my tool (requests from outside make UI not update correctly, sometimes flows seem "stopped" (that is not outputting to debug) while they, in reality work/are active, deploy button is enabled/disabled incorrectly etc. )
- by "node change" I understand (and tried): changing node position, changing node function code (insignificant from operation point of view like adding a comment), also tried significant change (inc counter by 10 instead of by 1) 

Actions I take:
- take clean node-red
- launch browser, go to UI
- import flow you gave in your post, click to place it
- do full deployment
- watch nice output with both counts in sync and timestamp
- change node outside the flow by adding line "//abc" as first line of it's function
- do deployment of changed nodes
- watch output with both counts still in sync and timestamp

same happens for change of function of node that is inside the subflow
BUT when I change the code to do "count+=10" instead of incrementing by 1, it increments by 10, starting from old count value, example:


```
51:51:<timestamp>
<-------------------- change nodes deploy here
52:61:<timestamp>
```

so the correct/new function gets executed.

As I see it strange here are the details of my setup:
```
12 Sep 20:00:21 - [info] Node-RED version: v0.13.4
12 Sep 20:00:21 - [info] Node.js  version: v4.4.4
12 Sep 20:00:21 - [info] Linux 4.2.0-42-generic x64 LE
```
I'll update tomorrow to latest node-red iff you think that may be the issue.

**Comment by knolleary** [comment]
Please do update to the latest version - lots of fixes will have gone in between 0.13.4 and the latest.

**Comment by knolleary** [comment]
Going to close this as I believe the original issue was clarified as a docs issue rather than a code one - and we've moved on a long way since 0.13.4.

Please reopen if you are able to reproduce on 0.16. Thanks!
